x-pack/legacy/plugins/lens/public/indexpattern_plugin/operations.ts

```diff
@@ -106,81 +104,89 @@ export function getOperationDisplay() {
   return display;
 }
 
-export function getOperationTypesForField(field: IndexPatternField): OperationType[] {
+export function getOperationTypesForField(field: IndexPatternField) {
   return operationDefinitions
-    .filter(definition => definition.isApplicableForField(field))
+    .filter(
+      operationDefinition => operationDefinition.getPossibleOperationsForField(field).length > 0
+    )
     .map(({ type }) => type);
 }
 
-export function buildColumnForOperationType<T extends OperationType>({
-  index,
+type OperationFieldTuple =
+  | { type: 'field'; operationType: OperationType; field: string }
+  | { operationType: OperationType; type: 'document' };
+
+export function getAvailableOperationsByMetaData(indexPattern: IndexPattern) {
+  const operationByMetaData: Record<string, OperationFieldTuple[]> = {};
```

Comment:
I found this a bit hard to follow, in part because the key was serving dual purposes. I think we can do away with the bit that requires us to later split the key. The split logic would be buggy if dataType ever had a '-' in it.

Here's the same code without requiring key to be parsed:

```js
export function getAvailableOperationsByMetaData(indexPattern) {
  const operationByMetaData = {};
  
  const addToMap = (operation) => (operationMetaData) => {
    const key = `${operationMetaData.dataType}-${operationMetaData.isBucketed ? 'true' : ''}`;

    if (operationByMetaData[key]) {
      operationByMetaData[key].operations.push(operation);
    } else {
      operationByMetaData[key] = {
        operationMetaData,
        operations: [operation],
      };
    }
  };

  operationDefinitions.forEach(operationDefinition => {
    operationDefinition
      .getPossibleOperationsForDocument(indexPattern)
      .forEach(addToMap({ type: 'document', operationType: operationDefinition.type }));

    indexPattern.fields.forEach(field => {
      operationDefinition.getPossibleOperationsForField(field).forEach(
        addToMap({
          type: 'field',
          operationType: operationDefinition.type,
          field: field.name,
        })
      );
    });
  });

  return Object.entries(operationByMetaData);
}
```

Reply:
Very good point, thanks.